Add declarative smart answer helper methods

kevindew opened the pull request:
Addresses #4510 - I've opened this up as a draft, as a proposal to resolve the issue

This adds `text_for`, `govspeak_for` and `html_for` methods intended
for the smart answer flow methods as a way for users to define content
for a smart answer. These are intended to replace the
`render_content_for` helper that has logic to determine the content
type.

This is set-up to be backwards compatible with the render_content_for
method so that both can co-exist for a period of time to allow
migration.

In doing this I also simplified a few things. We don't make any use of
options when calling `content_for` so these methods don't need to
support an option. All our usages of these methods use a block to add
content so I've set it up to raise an error.

theseanything commented:
Awesome 🎉 Thanks @kevindew - it does read rather nicely! I'd be happy with this change - what you've got in draft looks good. Think we just need some docs and tests for the new syntax. Then another PR or two to roll out the changes to the other smart answers?

Originally, I was worried about someone using the wrong helper method for content that can on be certain format (e.g. `govspeak_for :title`). Not sure how we'd mitigate that, think we are reliant on people noticing manually.

kevindew commented:
Thanks @theseanything I've made some updates for this now. I've updated the documentation, added tests for the functionality and I've also enabled the ability for smart answer flow files to have an extension of `.erb` so to not require the `.govspeak` part of the extension.

Regarding your concern of someone using `govspeak_for :title` style, I've iterated on the format definitions you introduced to set up a check for times where people use `govspeak_for` or `html_for`. Now if you do `govspeak_for :title` an exception will be raised. 

Previously I wasn't too concerned about that kind of mis-use, as it seems just one of many ways someone could make a mistake but it does seem like doing so could create a rather transparent issue. I've also realised that having this in will help with the migration over to this approach since it will break rather indignantly compared to succeeding quietly 😄

I think this is now ready for review. So I'll mark this so and open a draft of the big migration of answers 😬 .

theseanything commented on lib/smart_answer/erb_renderer/format_capture_helper.rb:
```diff
@@ -2,60 +2,80 @@ module SmartAnswer
+      :label,
+      :suffix_label,
+      /^error_/,
+    ].freeze
```
How come we switched to an array of mixed types? Feel this adds more complexity and reduces clarity of how these rules are evaluated.  E.g. what are allow types for this array, what takes precedence a symbol or regex, which one should I use?

kevindew replied:
It seemed easier to parse as non-regexs as you don't have to mentally subtract the `/^` and `$/` from each item to understand the contents. It also seems a relatively common pattern in Ruby to allow a regex for a pattern otherwise an exact match. Precedence doesn't impact as if any match it'll be deemed text only. I wasn't too concerned about the allowed types list since this a private concern of this class.

That said I'm not too fussed about it, just seemed a bummer to make all the names regexes for the sake of one that was a pattern. 

How strongly do you feel about it? prefer them all as regexs?

theseanything replied:
Yup! Sorry you're right, precedence (was my main concern) isn't an issue - was still thinking we are using these rules to distinguish between multiple format types.

theseanything commented on lib/smart_answer/erb_renderer/format_capture_helper.rb:
```diff
@@ -2,60 +2,80 @@ module SmartAnswer
+  private
+
+    def capture_content(&block)
+      raise "Expected a block" unless block
```
I assume we'll remove InvalidFormatType exception when we remove render_content_for, so we are okay with this inconsistency

theseanything commented on lib/smart_answer/erb_renderer/format_capture_helper.rb:
```diff
@@ -2,60 +2,80 @@ module SmartAnswer
+      end
+    end
+
+    def text_only_error(name)
```
Nit:
```suggestion
    def text_only_error_message(name)
```
The method name is confusing for the thing it is returning.

theseanything commented on doc/smart-answers/erb-templates.md:
```diff
@@ -1,27 +1,30 @@
 * [Outcome templates](erb-templates/outcome-templates.md)
 
-We remove all leading spaces from the content in the `render_content_for` blocks for Govspeak and Text formats. This allows us to indent the content in the `render_content_for` blocks without having to worry about it affecting the generated HTML when it's processed using Govspeak.
-
```
Do we convey the point that indentation is removed else where in the documentation? I might have missed it

kevindew replied:
I feel this is sufficiently implied through the examples. I didn't think this added a huge amount of value in this context. Do you think it's something people benefit from? could add something to individual functions.

theseanything replied:
Sorry defo not a blocker for this PR. I just anecdotally remember someone else finding this useful https://github.com/alphagov/smart-answers/pull/4502#discussion_r420162488

kevindew replied:
Argh sorry - I didn't realise it had been added so recently. I can understand that it raised an eyebrow for me to delete it.

kevindew replied:
I've added some info into the examples to say indentation is removed.

kevindew commented on doc/smart-answers/erb-templates.md:
```diff
@@ -1,27 +1,30 @@
 # ERB templates
 
-Content is defined in `render_content_for` blocks and with the name of the intent of the content:
+Content is defined using one of three helpers to indicate the format:
```
Sure, I've update the content here to try reflect your concern. It seemed a little bit tricky as it's not the helper that references the intent but the first argument of it.

kevindew commented on doc/smart-answers/erb-templates.md:
````diff
@@ -1,27 +1,30 @@
+```erb
+<% govspeak_for :body do %>
+  This uses the GOV.UK markdown dialect, govspeak, and will be converted to
+  HTML.
````
Good point 👍 I've also made this look more like govspeak to distinguish it from text.

kevindew commented on doc/smart-answers/erb-templates.md:
````diff
@@ -1,27 +1,30 @@
 ```
 
-We support Govspeak (`:govspeak`), plain text (`:text`) and HTML (`:html`). See docs for template pages for the default formats of different content blocks.
+For examples of which of these to use in which contexts see relevant documentation for different contexts:
````
Sure, I've applied your change here.

kevindew commented on doc/smart-answers/erb-templates/landing-page-template.md:
```diff
@@ -1,44 +1,53 @@
 
-* Default format is `:text`
-* Used as the heading (currently "h1")
+Used as the h1 heading, can only be text. Example:
```
Sorry about all of these, it was lazy editing from a previous version.

theseanything replied:
Np! Really minor anyway!

kevindew commented on doc/smart-answers/erb-templates/question-templates.md:
````diff
@@ -1,100 +1,114 @@
+<% end %>
+```
+
+Note: As of May 2020 this option may not be working correctly. Please check and amend this documentation.
````
Yup, that's a good idea. I'll do so.

kevindew replied:
I've opened https://github.com/alphagov/smart-answers/issues/4553

kevindew commented on doc/smart-answers/erb-templates/question-templates.md:
```diff
@@ -1,100 +1,114 @@
+[checkbox question](../question-types.md#checkbox_question) types. `hash`
+argument is a `Hash` of option keys (strings) and text values (also strings).
+It is used to "translate" options keys for multiple choice & checkbox questions
+into human-friendly text.
```
Yup - that makes sense. Muscle memory kicked in.

kevindew commented:
Thanks for the review @theseanything 👍 Something else has just come up so I've left this for now with just comments on your doc comments. Will return later. 

I've also removed the applying of this to the coronavirus example since it'll need rebasing and I think it's already served it's purpose as a demo.

theseanything commented:
@kevindew np!! No rush! - Feel free to hit the re-request review button 🛎️ when your ready!

kevindew commented on lib/smart_answer/erb_renderer.rb:
```diff
@@ -21,23 +23,16 @@ def option_text(key)
-
     def relative_erb_template_path
-      erb_template_path.relative_path_from(Rails.root).to_s
+      template_path = @view.lookup_context.find_template(@template_name).identifier
```
I've added an extra line to break this up.

kevindew commented on lib/smart_answer/erb_renderer/format_capture_helper.rb:
```diff
@@ -2,60 +2,80 @@ module SmartAnswer
+    def text_for(name, &block)
+      content = capture_content(&block).then { |c| strip_leading_spaces(c) }
+                                       .then { |c| normalize_blank_lines(c) }
+                                       .strip
```
Just that I think that using `then` to pipe a result is easier to follow than lots of assignments. I've switched it to a less fun version now though. 

I always seem to get push back in reviews when I use a `.then` 😞 it's one of my favourite recent Ruby additions.

theseanything replied:
Awww sorry! I actually don't mind the syntax, feels very JS-esque. Just thinking about style consistency for the rest of the codebase.

kevindew commented:
@theseanything this is good for another review - thanks 🐱
